File: desktop/android/app/src/main/java/org/pali/ime/PaliData.kt
// PaliData.kt — glossary/dictionary lookup + morphological splitter for the
// suggestion bar. Mirrors glossary.js + predict.js + PaliData.swift.
//
// Loads the curated bilingual glossary + roots/affixes (pali-data.json) and the
// full DPD dictionary (dpd-dict.json) from app assets. Pure logic (besides JSON
// loading) so analyze() can be unit-tested on the JVM.

package org.pali.ime

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

data class GlossResult(val en: String, val zh: String, val key: String, val stem: Boolean)
data class StemMatch(val kind: String, val label: String, val en: String, val zh: String) // root/word/raw
data class PrefixHit(val form: String, val en: String, val zh: String)
data class EndingHit(val end: String, val en: String, val zh: String)
data class Analysis(
    val prefixes: List<PrefixHit>,
    val stem: StemMatch,
    val ending: EndingHit?,
    val full: Boolean,
    val recognized: Int,
) {
    val morphemes: Int get() = prefixes.size + if (ending == null) 0 else 1
}

class PaliData private constructor(
    private val glossary: Map<String, Pair<String, String>>,
    private val dpd: JSONObject,
    private val rootForms: List<RootEntry>,
    private val prefixesAll: List<AffixEntry>,
    private val endingsSorted: List<EndingEntry>,
    private val glossAkk: List<GlossAkk>,
    private val freqWords: List<Triple<String, Int, String>>, // pre-sorted by freq desc
    private val compounds: Map<String, List<String>>,         // compound lemma -> members
) {
    private class RootEntry(val root: String, val en: String, val zh: String, val forms: List<Pair<String, List<String>>>)
    private class AffixEntry(val form: String, val en: String, val zh: String, val akk: List<String>)
    private class EndingEntry(val end: String, val en: String, val zh: String, val akk: List<String>)
    private class GlossAkk(val w: String, val akk: List<String>, val en: String, val zh: String)

    // MARK: gloss lookup (curated bilingual -> DPD English)
    fun lookup(word: String): GlossResult? {
        val k = word.lowercase()
        glossary[k]?.let { return GlossResult(it.first, it.second, k, false) }
        stemKey(k)?.let { k2 -> glossary[k2]?.let { return GlossResult(it.first, it.second, k2, true) } }
        if (dpd.has(k)) return GlossResult(dpd.optString(k), "", k, false)
        stemKey(k)?.let { k2 -> if (dpd.has(k2)) return GlossResult(dpd.optString(k2), "", k2, true) }
        // inflection fallback: analyze and return the best full stem's gloss
        val a = analyze(toAkk(k), 1).firstOrNull()
        if (a != null && a.full && a.stem.en.isNotEmpty()) {
            return GlossResult(a.stem.en, a.stem.zh, a.stem.label, true)
        }
        return null
    }
    private fun stemKey(k: String): String? =
        if (k.isNotEmpty() && (k.last() == 'ṃ' || k.last() == 'm')) k.dropLast(1) else null

    // MARK: compound (samāsa) split — member lemmas, or empty if not a compound
    fun splitCompound(word: String, lemma: String? = null): List<String> {
        val w = word.lowercase()
        compounds[w]?.let { return it }
        if (w.isNotEmpty() && (w.last() == 'ṃ' || w.last() == 'm')) compounds[w.dropLast(1)]?.let { return it }
        lemma?.lowercase()?.let { l -> compounds[l]?.let { return it } }
        return emptyList()
    }

    // MARK: word completion (frequency-ranked; freqWords is pre-sorted)
    fun completeWord(prefix: String, limit: Int = 6): List<Pair<String, String>> {
        val p = prefix.lowercase()
        if (p.isEmpty()) return emptyList()
        val out = ArrayList<Pair<String, String>>()
        for (t in freqWords) {
            if (t.first.length > p.length && t.first.startsWith(p)) {
                out.add(Pair(t.first, t.third))
                if (out.size >= limit) break
            }
        }
        return out
    }

    // MARK: akkhara helpers (lenient nasals / final-m)
    private fun akkEq(u: String, l: String): Boolean {
        if (u == l) return true
        if (u == "n" && (l == "ṅ" || l == "ñ" || l == "ṇ")) return true
        if (u == "m" && l == "ṃ") return true
        return false
    }
    private fun startsWith(lex: List<String>, pre: List<String>): Boolean {
        if (pre.size > lex.size) return false
        for (i in pre.indices) if (!akkEq(pre[i], lex[i])) return false
        return true
    }
    private fun endsWith(arr: List<String>, suf: List<String>): Boolean {
        if (suf.size > arr.size) return false
        val off = arr.size - suf.size
        for (i in suf.indices) if (!akkEq(arr[off + i], suf[i])) return false
        return true
    }
    private fun equalAkk(a: List<String>, b: List<String>) = a.size == b.size && startsWith(a, b)

    private fun matchStemWord(stem: List<String>): StemMatch? {
        for (g in glossAkk) if (equalAkk(g.akk, stem)) return StemMatch("word", g.w, g.en, g.zh)
        for (g in glossAkk) if (g.akk.size > stem.size && g.akk.size - stem.size <= 1 && startsWith(g.akk, stem))
            return StemMatch("word", g.w, g.en, g.zh)
        return null
    }
    private fun matchStemRoot(stem: List<String>): StemMatch? {
        for (rf in rootForms) for (f in rf.forms) {
            if (equalAkk(f.second, stem) ||
                (stem.size > f.second.size && stem.size - f.second.size <= 1 && startsWith(stem, f.second)))
                return StemMatch("root", "√" + rf.root, rf.en, rf.zh)
        }
        return null
    }
    private fun matchStemDpd(stem: List<String>): StemMatch? {
        val key = stem.joinToString("")
        if (dpd.has(key)) return StemMatch("word", key, dpd.optString(key), "")
        return null
    }

    // MARK: morphological split (prefix + root/word + ending)
    fun analyze(akk: List<String>, limit: Int = 2): List<Analysis> {
        if (akk.isEmpty()) return emptyList()

        val prefixOpts = ArrayList<Pair<List<PrefixHit>, List<String>>>()
        prefixOpts.add(Pair(emptyList(), akk))
        for (p in prefixesAll) {
            if (akk.size > p.akk.size && startsWith(akk, p.akk)) {
                val rest1 = akk.subList(p.akk.size, akk.size).toList()
                val ph = PrefixHit(p.form, p.en, p.zh)
                prefixOpts.add(Pair(listOf(ph), rest1))
                for (q in prefixesAll) {
                    if (rest1.size > q.akk.size && startsWith(rest1, q.akk)) {
                        prefixOpts.add(Pair(listOf(ph, PrefixHit(q.form, q.en, q.zh)), rest1.subList(q.akk.size, rest1.size).toList()))
                    }
                }
            }
        }

        val cands = ArrayList<Analysis>()
        for ((prefixes, rest) in prefixOpts) {
            val endOpts = ArrayList<Pair<EndingEntry?, List<String>>>()
            endOpts.add(Pair(null, rest))
            for (e in endingsSorted) {
                if (rest.size > e.akk.size && endsWith(rest, e.akk)) {
                    endOpts.add(Pair(e, rest.subList(0, rest.size - e.akk.size).toList()))
                }
            }
            for ((ending, stem) in endOpts) {
                if (stem.isEmpty()) continue
                val match = matchStemWord(stem) ?: matchStemRoot(stem) ?: matchStemDpd(stem)
                val stemM = match ?: StemMatch("raw", stem.joinToString(""), "", "")
                val pfxLen = akk.size - rest.size
                val endLen = rest.size - stem.size
                val stemLen = if (match != null) stem.size else 0
                val endingHit = ending?.let { EndingHit(it.end, it.en, it.zh) }
                cands.add(Analysis(prefixes, stemM, endingHit, match != null, pfxLen + endLen + stemLen))
            }
        }

        cands.sortWith(
            compareByDescending<Analysis> { it.full }
                .thenByDescending { it.recognized }
                .thenBy { it.morphemes }
                .thenBy { it.prefixes.size }
        )
        val seen = HashSet<String>()
        val dedup = ArrayList<Analysis>()
        for (c in cands) {
            val sig = c.prefixes.joinToString("+") { it.form } + "|" + c.stem.label
            if (seen.add(sig)) dedup.add(c)
        }
        val fulls = dedup.filter { it.full }
        return if (fulls.isNotEmpty()) fulls.take(limit) else dedup.take(1)
    }

    companion object {
        fun toAkk(s: String): List<String> =
            PaliEngine.tokenize(s, false).filter { it.type != TokType.OTHER }.map { it.iast }

        fun load(ctx: Context): PaliData {
            val pt = ctx.assets.open("pali-data.json").bufferedReader().use { it.readText() }
            val dt = ctx.assets.open("dpd-dict.json").bufferedReader().use { it.readText() }
            val ft = ctx.assets.open("freq-words.json").bufferedReader().use { it.readText() }
            val ct = ctx.assets.open("compounds.json").bufferedReader().use { it.readText() }
            return build(JSONObject(pt), JSONObject(dt), ft, ct)
        }
        // For JVM tests: construct from JSON text directly.
        fun fromJson(paliText: String, dpdText: String, freqText: String, compoundsText: String): PaliData =
            build(JSONObject(paliText), JSONObject(dpdText), freqText, compoundsText)

        private fun build(pali: JSONObject, dpd: JSONObject, freqText: String, compoundsText: String): PaliData {
            val gmap = HashMap<String, Pair<String, String>>()
            val gloss = pali.getJSONObject("glossary")
            for (k in gloss.keys()) {
                val o = gloss.getJSONObject(k)
                gmap[k] = Pair(o.optString("en"), o.optString("zh"))
            }
            val glossAkk = gmap.map { (k, v) -> GlossAkk(k, toAkk(k), v.first, v.second) }

            val rootForms = ArrayList<RootEntry>()
            val ra = pali.getJSONArray("roots")
            for (i in 0 until ra.length()) {
                val o = ra.getJSONObject(i)
                val fa = o.getJSONArray("forms")
                val forms = ArrayList<Pair<String, List<String>>>()
                for (j in 0 until fa.length()) { val f = fa.getString(j); forms.add(Pair(f, toAkk(f))) }
                rootForms.add(RootEntry(o.optString("root"), o.optString("en"), o.optString("zh"), forms))
            }

            val prefixesAll = ArrayList<AffixEntry>()
            for (key in listOf("upasagga", "prefixExtra")) {
                val arr = pali.optJSONArray(key) ?: continue
                for (i in 0 until arr.length()) {
                    val o = arr.getJSONObject(i)
                    prefixesAll.add(AffixEntry(o.optString("form"), o.optString("en"), o.optString("zh"), toAkk(o.optString("form"))))
                }
            }

            val endings = ArrayList<EndingEntry>()
            val ea = pali.getJSONArray("endings")
            for (i in 0 until ea.length()) {
                val o = ea.getJSONObject(i)
                endings.add(EndingEntry(o.optString("end"), o.optString("en"), o.optString("zh"), toAkk(o.optString("end"))))
            }
            endings.sortByDescending { it.akk.size }

            val freqWords = ArrayList<Triple<String, Int, String>>()
            val fa = JSONArray(freqText)
            for (i in 0 until fa.length()) {
                val row = fa.getJSONArray(i)
                freqWords.add(Triple(row.getString(0), row.optInt(1, 0), if (row.length() >= 3) row.optString(2) else ""))
            }

            val comp = HashMap<String, List<String>>()
            val cj = JSONObject(compoundsText)
            for (k in cj.keys()) {
                val arr = cj.getJSONArray(k)
                val ms = ArrayList<String>(arr.length())
                for (i in 0 until arr.length()) ms.add(arr.getString(i))
                comp[k] = ms
            }

            return PaliData(gmap, dpd, rootForms, prefixesAll, endings, glossAkk, freqWords, comp)
        }
    }
}

File: desktop/macos/Sources/PaliData.swift
// PaliData.swift — loads pali-data.json (glossary + roots/affixes exported from
// the web app) and ports the glossary lookup + morphological splitter
// (mirrors glossary.js + predict.js). Used by the IME's info panel to show
// meanings and prefix+root+ending analysis as you type.

import Foundation

struct Gloss: Codable { let en: String; let zh: String }
struct Root: Codable { let root: String; let forms: [String]; let en: String; let zh: String }
struct Affix: Codable { let form: String; let en: String; let zh: String }
struct Ending: Codable { let end: String; let kind: String; let en: String; let zh: String }

private struct Bundled: Codable {
    let glossary: [String: Gloss]
    let phrases: [String: Gloss]
    let roots: [Root]
    let upasagga: [Affix]
    let endings: [Ending]
    let prefixExtra: [Affix]
}

// Lookup / analysis results
struct GlossResult { let en: String; let zh: String; let key: String; let stem: Bool }
struct Completion { let w: String; let en: String }
struct StemMatch { let kind: String; let label: String; let en: String; let zh: String } // kind: root/word/raw
struct PrefixHit { let form: String; let en: String; let zh: String }
struct EndingHit { let end: String; let en: String; let zh: String }
struct Analysis {
    let prefixes: [PrefixHit]
    let stem: StemMatch
    let ending: EndingHit?
    let full: Bool
    let recognized: Int
    var morphemes: Int { prefixes.count + (ending == nil ? 0 : 1) }
}

final class PaliData {
    static let shared: PaliData? = {
        guard let url = Bundle.main.url(forResource: "pali-data", withExtension: "json") else { return nil }
        let dpdURL = Bundle.main.url(forResource: "dpd-dict", withExtension: "json")
        let freqURL = Bundle.main.url(forResource: "freq-words", withExtension: "json")
        let compURL = Bundle.main.url(forResource: "compounds", withExtension: "json")
        let bigramURL = Bundle.main.url(forResource: "bigram", withExtension: "json")
        return PaliData(url: url, dpdURL: dpdURL, freqURL: freqURL, compoundsURL: compURL, bigramURL: bigramURL)
    }()

    private let data: Bundled
    private let dpd: [String: String]   // full DPD lemma -> English meaning (75k+)
    private let freqWords: [(w: String, en: String)]  // frequency-ranked, for completion
    private let compounds: [String: [String]]  // compound lemma -> member lemmas
    private let bigram: [String: [String]]     // word -> top next words (corpus order)
    private let rootForms: [(root: Root, forms: [(form: String, akk: [String])])]
    private let prefixesAll: [(aff: Affix, akk: [String])]
    private let endingsSorted: [(end: Ending, akk: [String])]
    private let glossAkk: [(w: String, akk: [String], en: String, zh: String)]

    init?(url: URL, dpdURL: URL? = nil, freqURL: URL? = nil, compoundsURL: URL? = nil, bigramURL: URL? = nil) {
        guard let raw = try? Data(contentsOf: url),
              let d = try? JSONDecoder().decode(Bundled.self, from: raw) else { return nil }
        data = d
        if let cu = compoundsURL, let craw = try? Data(contentsOf: cu),
           let cc = try? JSONDecoder().decode([String: [String]].self, from: craw) {
            compounds = cc
        } else {
            compounds = [:]
        }
        if let bu = bigramURL, let braw = try? Data(contentsOf: bu),
           let bb = try? JSONDecoder().decode([String: [String]].self, from: braw) {
            bigram = bb
        } else {
            bigram = [:]
        }
        if let du = dpdURL, let draw = try? Data(contentsOf: du),
           let dd = try? JSONDecoder().decode([String: String].self, from: draw) {
            dpd = dd
        } else {
            dpd = [:]
        }
        // freq-words.json is an array of [lemma, freq, meaning], pre-sorted by
        // frequency; parsed loosely (heterogeneous arrays) via JSONSerialization.
        if let fu = freqURL, let fraw = try? Data(contentsOf: fu),
           let arr = try? JSONSerialization.jsonObject(with: fraw) as? [[Any]] {
            freqWords = arr.compactMap { row in
                guard let w = row.first as? String else { return nil }
                let en = row.count >= 3 ? (row[2] as? String ?? "") : ""
                return (w, en)
            }
        } else {
            freqWords = []
        }
        func toA(_ s: String) -> [String] {
            PaliEngine.tokenize(s, smartNasal: false).filter { $0.type != .other }.map { $0.iast }
        }
        rootForms = d.roots.map { r in (r, r.forms.map { (form: $0, akk: toA($0)) }) }
        prefixesAll = (d.upasagga + d.prefixExtra).map { ($0, toA($0.form)) }
        endingsSorted = d.endings.map { ($0, toA($0.end)) }.sorted { $0.1.count > $1.1.count }
        glossAkk = d.glossary.map { (w: $0.key, akk: toA($0.key), en: $0.value.en, zh: $0.value.zh) }
    }

    // MARK: akkhara helpers (lenient nasals / final-m, mirroring predict.js)
    func toAkk(_ iast: String) -> [String] {
        PaliEngine.tokenize(iast, smartNasal: false).filter { $0.type != .other }.map { $0.iast }
    }
    private func akkEq(_ u: String, _ l: String) -> Bool {
        if u == l { return true }
        if u == "n" && (l == "ṅ" || l == "ñ" || l == "ṇ") { return true }
        if u == "m" && l == "ṃ" { return true }
        return false
    }
    private func startsWith(_ lex: [String], _ pre: [String]) -> Bool {
        if pre.count > lex.count { return false }
        for i in 0..<pre.count where !akkEq(pre[i], lex[i]) { return false }
        return true
    }
    private func endsWith(_ arr: [String], _ suf: [String]) -> Bool {
        if suf.count > arr.count { return false }
        let off = arr.count - suf.count
        for i in 0..<suf.count where !akkEq(arr[off + i], suf[i]) { return false }
        return true
    }
    private func equalAkk(_ a: [String], _ b: [String]) -> Bool {
        a.count == b.count && startsWith(a, b)
    }

    // MARK: glossary lookup (exact, then strip a trailing ṃ/m)
    func lookup(_ iastWord: String) -> GlossResult? {
        let k = iastWord.lowercased()
        // 1) curated bilingual glossary (exact, then strip a trailing ṃ/m)
        if let g = data.glossary[k] { return GlossResult(en: g.en, zh: g.zh, key: k, stem: false) }
        if let last = k.last, last == "ṃ" || last == "m" {
            let k2 = String(k.dropLast())
            if let g = data.glossary[k2] { return GlossResult(en: g.en, zh: g.zh, key: k2, stem: true) }
        }
        // 2) full DPD dictionary (English only)
        if let m = dpd[k] { return GlossResult(en: m, zh: "", key: k, stem: false) }
        if let last = k.last, last == "ṃ" || last == "m" {
            let k2 = String(k.dropLast())
            if let m = dpd[k2] { return GlossResult(en: m, zh: "", key: k2, stem: true) }
        }
        // 3) inflection fallback: morphologically analyze and return the best
        //    fully-explained stem's gloss (so declined/conjugated forms resolve).
        if let a = analyze(toAkk(k), limit: 1).first, a.full, !a.stem.en.isEmpty {
            return GlossResult(en: a.stem.en, zh: a.stem.zh, key: a.stem.label, stem: true)
        }
        return nil
    }

    // MARK: compound (samāsa) split — member lemmas, or [] if not a compound
    func splitCompound(_ word: String, lemma: String? = nil) -> [String] {
        let w = word.lowercased()
        if let m = compounds[w] { return m }
        if let last = w.last, last == "ṃ" || last == "m", let m = compounds[String(w.dropLast())] { return m }
        if let l = lemma?.lowercased(), let m = compounds[l] { return m }
        return []
    }

    // MARK: word completion (frequency-ranked; freqWords is pre-sorted)
    func completeWord(_ prefix: String, limit: Int = 6) -> [Completion] {
        let p = prefix.lowercased()
        if p.isEmpty { return [] }
        var out: [Completion] = []
        for e in freqWords where e.w.count > p.count && e.w.hasPrefix(p) {
            out.append(Completion(w: e.w, en: e.en))
            if out.count >= limit { break }
        }
        return out
    }

    // MARK: next-word prediction (bigram model from the Pali canon)
    // Given the word just committed, return likely following words (with glosses).
    func nextWord(_ word: String, limit: Int = 5) -> [Completion] {
        let w = word.lowercased()
        var succ = bigram[w]
        if succ == nil, let last = w.last, last == "ṃ" || last == "m" {
            succ = bigram[String(w.dropLast())]
        }
        guard let words = succ else { return [] }
        return words.prefix(limit).map { Completion(w: $0, en: lookup($0)?.en ?? "") }
    }

    // MARK: morphological split (prefix + root/word + ending)
    private func matchStemWord(_ stem: [String]) -> StemMatch? {
        for g in glossAkk where equalAkk(g.akk, stem) {
            return StemMatch(kind: "word", label: g.w, en: g.en, zh: g.zh)
        }
        for g in glossAkk where g.akk.count > stem.count && g.akk.count - stem.count <= 1 && startsWith(g.akk, stem) {
            return StemMatch(kind: "word", label: g.w, en: g.en, zh: g.zh)
        }
        return nil
    }
    // Last resort: the full DPD dictionary (exact stem, English only). Tried
    // after roots so the morphological panel prefers the more informative √root.
    private func matchStemDpd(_ stem: [String]) -> StemMatch? {
        let key = stem.joined()
        if let m = dpd[key] { return StemMatch(kind: "word", label: key, en: m, zh: "") }
        return nil
    }
    private func matchStemRoot(_ stem: [String]) -> StemMatch? {
        for rf in rootForms {
            for f in rf.forms {
                if equalAkk(f.akk, stem) ||
                    (stem.count > f.akk.count && stem.count - f.akk.count <= 1 && startsWith(stem, f.akk)) {
                    return StemMatch(kind: "root", label: "√" + rf.root.root, en: rf.root.en, zh: rf.root.zh)
                }
            }
        }
        return nil
    }

    func analyze(_ akk: [String], limit: Int = 2) -> [Analysis] {
        if akk.isEmpty { return [] }

        var prefixOpts: [(prefixes: [PrefixHit], rest: [String])] = [([], akk)]
        for p in prefixesAll where akk.count > p.akk.count && startsWith(akk, p.akk) {
            let rest1 = Array(akk[p.akk.count...])
            let ph = PrefixHit(form: p.aff.form, en: p.aff.en, zh: p.aff.zh)
            prefixOpts.append(([ph], rest1))
            for q in prefixesAll where rest1.count > q.akk.count && startsWith(rest1, q.akk) {
                let qh = PrefixHit(form: q.aff.form, en: q.aff.en, zh: q.aff.zh)
                prefixOpts.append(([ph, qh], Array(rest1[q.akk.count...])))
            }
        }

        var cands: [Analysis] = []
        for po in prefixOpts {
            var endOpts: [(ending: Ending?, stem: [String])] = [(nil, po.rest)]
            for e in endingsSorted where po.rest.count > e.akk.count && endsWith(po.rest, e.akk) {
                endOpts.append((e.end, Array(po.rest[0..<(po.rest.count - e.akk.count)])))
            }
            for eo in endOpts {
                if eo.stem.isEmpty { continue }
                let match = matchStemWord(eo.stem) ?? matchStemRoot(eo.stem) ?? matchStemDpd(eo.stem)
                let stem = match ?? StemMatch(kind: "raw", label: eo.stem.joined(), en: "", zh: "")
                let pfxLen = akk.count - po.rest.count
                let endLen = po.rest.count - eo.stem.count
                let stemLen = match != nil ? eo.stem.count : 0
                let endingHit = eo.ending.map { EndingHit(end: $0.end, en: $0.en, zh: $0.zh) }
                cands.append(Analysis(prefixes: po.prefixes, stem: stem, ending: endingHit,
                                      full: match != nil, recognized: pfxLen + endLen + stemLen))
            }
        }

        cands.sort {
            if $0.full != $1.full { return $0.full && !$1.full }
            if $0.recognized != $1.recognized { return $0.recognized > $1.recognized }
            if $0.morphemes != $1.morphemes { return $0.morphemes < $1.morphemes }
            return $0.prefixes.count < $1.prefixes.count
        }
        var seen = Set<String>()
        var dedup: [Analysis] = []
        for c in cands {
            let sig = c.prefixes.map { $0.form }.joined(separator: "+") + "|" + c.stem.label
            if seen.contains(sig) { continue }
            seen.insert(sig)
            dedup.append(c)
        }
        let fulls = dedup.filter { $0.full }
        return fulls.isEmpty ? Array(dedup.prefix(1)) : Array(fulls.prefix(limit))
    }
}
